File: frontend-admin/src/types/notification.ts
/**
 * Phase 5 站内信 / 消息通知类型（Admin 视角）。
 *
 * 与 docs/API/phase-5-contracts.md §3.4 / §6 严格对齐。
 *
 * Admin 端 scope 前缀为 `/admin`；其它端结构一致。
 */

// ---------------------------------------------------------------------------
// 枚举
// ---------------------------------------------------------------------------

/** 契约 §3.4 recipient_type。Admin 收件人仅可能为 `admin`。 */
export type NotificationRecipientType = "user" | "merchant" | "admin";

/** 契约 §3.4 category（6 类）。 */
export type NotificationCategory =
  | "system"
  | "order"
  | "aftersales"
  | "review"
  | "shop"
  | "promo";

// ---------------------------------------------------------------------------
// 数据结构
// ---------------------------------------------------------------------------

/**
 * 站内信 / 通知消息（契约 §3.4）。
 *
 * 展示注意：
 * - body 支持简单 markdown（Phase 5 前端按纯文本 + 换行渲染即可）
 * - action_url 为相对路径（例如 `/console/aftersales/123`），前端点击跳转
 */
export interface NotificationOut {
  id: number;
  recipient_type: NotificationRecipientType;
  recipient_id: number;
  category: NotificationCategory;
  title: string;
  body: string;
  action_url: string | null;
  related_type: string | null;
  related_id: number | null;
  is_read: boolean;
  read_at: string | null;
  created_at: string;
}

// ---------------------------------------------------------------------------
// 查询 / 响应负载
// ---------------------------------------------------------------------------

/** GET /admin/notifications 查询参数。 */
export interface ListNotificationsQuery {
  is_read?: boolean;
  category?: NotificationCategory;
  page?: number;
  size?: number;
}

/** GET /admin/notifications/unread-count 响应体。 */
export interface UnreadCountData {
  count: number;
}

File: backend/app/models/notification.py
"""Notification ORM model — Phase 5 contract §3.4.

Cross-role in-app notifications for user / merchant / admin. Delivery
is best-effort HTTP polling (WebSocket deferred). No soft-delete —
"clear read" is a hard DELETE.
"""

from __future__ import annotations

import enum
from datetime import datetime

from sqlalchemy import DateTime, Enum, Index, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, BigIntId, IdMixin, TimestampMixin


class NotificationRecipientType(enum.StrEnum):
    """Who receives the notification."""

    USER = "user"
    MERCHANT = "merchant"
    ADMIN = "admin"


class NotificationCategory(enum.StrEnum):
    """High-level bucket for filtering in the inbox UI."""

    SYSTEM = "system"
    ORDER = "order"
    AFTERSALES = "aftersales"
    REVIEW = "review"
    SHOP = "shop"
    PROMO = "promo"


class Notification(IdMixin, TimestampMixin, Base):
    """A single in-app notification row."""

    __tablename__ = "notifications"

    recipient_type: Mapped[NotificationRecipientType] = mapped_column(
        Enum(
            NotificationRecipientType,
            name="notification_recipient_type",
            native_enum=True,
            validate_strings=True,
        ),
        nullable=False,
    )
    recipient_id: Mapped[int] = mapped_column(BigIntId, nullable=False)
    category: Mapped[NotificationCategory] = mapped_column(
        Enum(
            NotificationCategory,
            name="notification_category",
            native_enum=True,
            validate_strings=True,
        ),
        nullable=False,
    )
    title: Mapped[str] = mapped_column(String(120), nullable=False)
    body: Mapped[str] = mapped_column(Text, nullable=False)
    action_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    related_type: Mapped[str | None] = mapped_column(String(60), nullable=True)
    related_id: Mapped[int | None] = mapped_column(BigIntId, nullable=True)

    is_read: Mapped[bool] = mapped_column(nullable=False, default=False)
    read_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    __table_args__ = (
        Index(
            "ix_notifications_inbox",
            "recipient_type",
            "recipient_id",
            "is_read",
            "created_at",
        ),
    )
